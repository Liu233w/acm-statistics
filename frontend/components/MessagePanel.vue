<template>
  <v-expand-transition>
    <v-card
      v-show="show"
      class="message-panel"
    >
      <v-slide-y-transition
        group
        tag="v-list"
        class="pa-0"
      >
        <v-list-item
          v-for="(item,i) in list"
          :key="item.id"
          class="pa-0"
        >
          <v-alert
            dismissible
            :dense="!oneItem"
            :type="item.type"
            @input="closeOne(i)"
            :class="{ 'mb-0': oneItem, 'mb-1': !oneItem }"
          >
            {{ item.message }}
          </v-alert>
        </v-list-item>
      </v-slide-y-transition>
      <v-card-actions
        v-if="!oneItem"
        class="pt-0"
      >
        <v-spacer />
        <v-btn
          text
          color="primary"
          @click="close"
        >
          close
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-expand-transition>
</template>

<script>
import { mapGetters, mapState } from 'vuex'

export default {

  computed: {
    ...mapState('message', [
      'list',
    ]),
    ...mapGetters('message', [
      'top',
    ]),
    show() {
      return this.$store.getters['message/show']
    },
    oneItem() {
      return this.list.length <= 1
    },
  },

  methods: {
    close() {
      this.$store.commit('message/clear')
    },
    closeOne(i) {
      this.$store.commit('message/remove', i)
    },
  },
}
</script>

<style scoped>
.message-panel {
  position: fixed;
  z-index: 1000;
  top: 10px;

  left: 50%;
  transform: translateX(-50%);

  transition: flex 0.3s ease-out;
}
</style>
