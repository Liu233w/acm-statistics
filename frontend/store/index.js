export const state = () => ({
  sidebar: false,
})

export const mutations = {
  toggleSidebar (state) {
    state.sidebar = !state.sidebar
  },
}

export const actions = {
  async nuxtServerInit({ dispatch }) {
    await dispatch('session/refreshUser')
  },
}
