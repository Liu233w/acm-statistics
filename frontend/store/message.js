import _ from 'lodash'

export const namespaced = true

export const state = () => ({
  list: [],
})

function addMessage(state, message, type) {
  state.list.push({
    id: Math.random(),
    message,
    type,
  })
}

export const mutations = {
  add(state, param) {
    if (_.isString(param)) {
      addMessage(state, param, 'info')
    } else {
      const { message, type } = param
      addMessage(state, message, type)
    }
  },
  pop(state) {
    state.list.pop()
  },
  addError(state, message) {
    addMessage(state, message, 'error')
  },
  addSuccess(state, message) {
    addMessage(state, message, 'success')
  },
  remove(state, index) {
    state.list.splice(index, 1)
  },
  clear(state) {
    state.list = []
  },
}

export const getters = {
  top(state) {
    if (state.list.length > 0) {
      return state.list[state.list.length - 1]
    } else {
      return {}
    }
  },
  show(state) {
    return state.list.length > 0
  },
}
