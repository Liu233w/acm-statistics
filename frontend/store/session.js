import Cookies from 'js-cookie'
import _ from 'lodash'

export const state = () => ({
  login: false,
  username: '',
  settings: {},
})

export const mutations = {
  setUser(state, { username }) {
    state.login = true
    state.username = username
  },
  removeUser(state) {
    state.login = false
    state.username = ''
  },
  updateSettings(state, newSettings) {
    state.settings = _.assign(state.settings, newSettings)
  },
  resetSettings(state) {
    state.settings = {}
  },
}

export const actions = {
  async refreshUser({ commit, dispatch }) {
    try {
      const res = await this.$axios.$get('/api/services/app/Session/GetCurrentLoginInformations')

      if (res.result.user) {
        commit('setUser', { username: res.result.user.userName })
        await dispatch('refreshSettings')
      } else {
        commit('removeUser')
        commit('resetSettings')
      }
    } catch (err) {
      Cookies.remove('OAuthToken')
    }
  },
  async login({ dispatch }, { username, password, remember }) {
    const res = await this.$axios.$post('/api/TokenAuth/Authenticate', {
      userNameOrEmailAddress: username,
      password: password,
      rememberClient: remember,
    })

    const config = {}
    if (remember) {
      config.expires = 30
    }
    Cookies.set('OAuthToken', res.result.accessToken, config)

    await dispatch('refreshUser')
    await dispatch('refreshSettings')
  },
  async logout({ commit }) {
    Cookies.remove('OAuthToken')
    commit('removeUser')
    commit('resetSettings')
  },
  async refreshSettings({ commit }) {
    const res = await this.$axios.$get('/api/services/app/UserConfig/GetUserSettings')
    commit('updateSettings', res.result)
  },
}
