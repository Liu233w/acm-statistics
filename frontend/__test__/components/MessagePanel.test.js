import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import Vuetify from 'vuetify'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuetify)

import MessagePanel from '~/components/MessagePanel.vue'
import * as MessageStore from '~/store/message'

describe('MessagePanel', () => {

  let wrapper
  let store

  beforeEach(() => {
    const vuetify = new Vuetify()
    store = new Vuex.Store({
      modules: {
        message: MessageStore,
      },
    })
    wrapper = mount(MessagePanel, {
      store,
      localVue,
      vuetify,
    })
  })

  it('can render correctly', async () => {
    expect(wrapper.vm).toBeTruthy()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can render when there are 1 element', async () => {
    store.commit('message/add', 'Info')
    await wrapper.vm.$nextTick()
    expect(store.state.message.list).toHaveLength(1)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can render when there are 2 elements', async () => {
    store.commit('message/add', {
      message: 'Info',
      type: 'info',
    })
    store.commit('message/add', {
      message: 'Error',
      type: 'error',
    })
    await wrapper.vm.$nextTick()
    expect(store.state.message.list).toHaveLength(2)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
