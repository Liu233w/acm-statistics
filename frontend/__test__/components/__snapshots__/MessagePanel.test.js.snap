// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagePanel can render correctly 1`] = `
<transition-stub name="expand-transition">
  <div class="message-panel v-card v-sheet theme--light" style="display: none;">
    <transition-group-stub tag="v-list" name="slide-y-transition" class="pa-0"></transition-group-stub>
    <!---->
  </div>
</transition-stub>
`;

exports[`MessagePanel can render when there are 1 element 1`] = `
<transition-stub name="expand-transition">
  <div class="message-panel v-card v-sheet theme--light" style="">
    <transition-group-stub tag="v-list" name="slide-y-transition" class="pa-0">
      <div tabindex="-1" class="pa-0 v-list-item theme--light">
        <div role="alert" class="v-alert v-sheet theme--dark info mb-0">
          <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon material-icons theme--dark">$info</i>
            <div class="v-alert__content">
              Info
            </div><button type="button" class="v-alert__dismissible v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--small" aria-label="Close"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">$cancel</i></span></button>
          </div>
        </div>
      </div>
    </transition-group-stub>
    <!---->
  </div>
</transition-stub>
`;

exports[`MessagePanel can render when there are 2 elements 1`] = `
<transition-stub name="expand-transition">
  <div class="message-panel v-card v-sheet theme--light" style="">
    <transition-group-stub tag="v-list" name="slide-y-transition" class="pa-0">
      <div tabindex="-1" class="pa-0 v-list-item theme--light">
        <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense info mb-1">
          <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon material-icons theme--dark">$info</i>
            <div class="v-alert__content">
              Info
            </div><button type="button" class="v-alert__dismissible v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--small" aria-label="Close"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">$cancel</i></span></button>
          </div>
        </div>
      </div>
      <div tabindex="-1" class="pa-0 v-list-item theme--light">
        <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense error mb-1">
          <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon material-icons theme--dark">$error</i>
            <div class="v-alert__content">
              Error
            </div><button type="button" class="v-alert__dismissible v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--small" aria-label="Close"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">$cancel</i></span></button>
          </div>
        </div>
      </div>
    </transition-group-stub>
    <div class="v-card__actions pt-0">
      <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">
        close
      </span></button>
    </div>
  </div>
</transition-stub>
`;
